--- zstd-1.5.0/build/meson/tests/meson.build	2021-05-14 22:59:34.000000000 +0800
+++ zstd.bak/build/meson/tests/meson.build	2021-11-08 17:12:17.225379241 +0800
@@ -186,10 +186,6 @@
   zstreamtest,
   args: ['-v', ZSTREAM_TESTTIME] + FUZZER_FLAGS,
   timeout: 240)
-test('test-zstream-2',
-  zstreamtest,
-  args: ['-mt', '-t1', ZSTREAM_TESTTIME] + FUZZER_FLAGS,
-  timeout: 120)
 test('test-zstream-3',
   zstreamtest,
   args: ['--newapi', '-t1', ZSTREAM_TESTTIME] + FUZZER_FLAGS,
