diff -Nur zstd-1.5.0/tests/playTests.sh new-zstd/tests/playTests.sh
--- zstd-1.5.0/tests/playTests.sh	2021-05-14 22:59:34.000000000 +0800
+++ new-zstd/tests/playTests.sh	2021-11-18 16:44:28.963975598 +0800
@@ -689,6 +689,10 @@
 test -f tmp3
 test -f tmp4
 
+println "test : survive a list of files which long file name length (--filelist=FILE)"
+datagen -g51M > tmp_badList 
+zstd -f --filelist=tmp_badList && die "should have failed : list file name length is too long" # can trigger the file list file name length is too long
+
 println "test : survive a list of files which is text garbage (--filelist=FILE)"
 datagen > tmp_badList
 zstd -f --filelist=tmp_badList && die "should have failed : list is text garbage"
